certora/specs/DoubleEndedQueue.spec

```diff
@@ -0,0 +1,408 @@
+
+    assert afterAt == beforeAt, "data is preserved";
+}
+
+/*
+┌─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
+│ Function correctness: pushBack adds an element at the end of the queue                                              │
+└─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
+*/
+rule pushBack {
+    require boundedQueue();
+
+    bytes32 value;
+
+    bytes32 backBefore = back();
+
+    pushBack@withrevert(value);
+    bool success = !lastReverted;
+
+    bytes32 backAfter = back();
+
+    // liveness
+    assert success, "never reverts";
+
+    // effect
+    assert success => backAfter == value, "back set to value";
+
+    // no side effect
+    assert backBefore == backAfter <=> backBefore == value, "back doesn't change only if it was already set";
+}
+
+/*
+┌─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
+│ Rule: pushBack preserves the previous values in the queue                                                           │
+└─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
+*/
+rule pushBackConsistency {
+    require boundedQueue();
+
+    uint256 key;
+    bytes32 beforeAt = at_(key);
+
+    bytes32 value;
+    pushBack@withrevert(value);
+
+    bytes32 afterAt = at_(key);
+
+    assert afterAt == beforeAt, "data is preserved";
+}
+
+/*
+┌─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
+│ Function correctness: popFront removes an element from the beginning of the queue                                   │
+└─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
+*/
+rule popFront {
+    requireInvariant boundariesConsistency();
+    require boundedQueue();
+
+    bool emptyBefore = empty();
+    bytes32 frontBefore = front@withrevert();
+
+    bytes32 popped = popFront@withrevert();
+    bool success = !lastReverted;
+
+    // liveness
+    assert success <=> !emptyBefore, "only fails if it's empty";
+
+    // effect
+    assert success => frontBefore == popped, "previous front is returned";
+}
+
+/*
+┌─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
+│ Rule: front is previous at(1) after calling popFront                                                                │
+└─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
+*/
+rule popFrontExpected {
+    requireInvariant boundariesConsistency();
+    require boundedQueue();
+    
+    bytes32 nextFront = at_(1);
+    popFront@withrevert();
+
+    assert !lastReverted <=> front() == nextFront, "front is adjusted";
+}
```

Comment:
This is a consequence of `popFrontConsistency` + `queueEndings`. I'd remove that rule.

Reply:
That's correct, thanks!